Add recipe for consult-gh-emacs-pr-review (renamed to consult-gh-with-pr-review)

armindarvish opened the pull request:
### Brief summary of what the package does

Adds pr-review integration for consult-gh  
  


### Direct link to the package repository

<https://github.com/armindarvish/consult-gh>  
  


### Your association with the package

I am the maintainer.  
  


### Relevant communications with the upstream package maintainer

None needed  
  


### Checklist

-   The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
-   I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
-   I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
-   My elisp byte-compiles cleanly
-   I've used `M-x checkdoc` to check the package's documentation strings
-   I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

armindarvish commented:
I fixed the CI errors.

riscy commented:
Thanks for this.  Quick first pass -

⸺ `consult-gh-emacs-pr-review.el` with byte-compile using Emacs 29.4:
```
consult-gh-emacs-pr-review.el:37:2: Error: Cannot open load file: No such file or directory, consult-gh
```

⸺ `consult-gh-emacs-pr-review.el` with package-lint 20250108.1407:
```
4 issues found:
1:0: warning: The word "emacs" is redundant in Emacs package names.
9:23: warning: This makes the package uninstallable in all released Emacs versions.
9:38: warning: Use a non-snapshot version number for dependency on "consult" if possible.
138:0: error: "consult-gh-topics-open-in-emacs-pr-review" doesn't start with package's prefix "consult-gh-emacs-pr-review".
```

⸺ `consult-gh-emacs-pr-review.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- consult-gh-emacs-pr-review.el#L150: No `format` required; `message` takes an f-string
```

⸺ `consult-gh-emacs-pr-review.el` with checkdoc 0.6.2 (fix *within reason*):
```
consult-gh-emacs-pr-review.el:42: Lisp symbol `pr-review' should appear in quotes
consult-gh-emacs-pr-review.el:52: Lisp symbol `pr-review' should appear in quotes
consult-gh-emacs-pr-review.el:139: Lisp symbol `pr-review' should appear in quotes
```

armindarvish commented:
> 1:0: warning: The word "emacs" is redundant in Emacs package names  

In this case I am calling functions, variables etc. with the prefix `consult-gh-emacs-pr-review` to indicate intergation with emacs-pr-review (pr-review) package. This avoids confusion with `consult-gh-pr-review`, which is already used in consult-gh for the pr-review in consult-gh itself.  

> ⸺ `consult-gh-emacs-pr-review.el` with byte-compile using Emacs 29.4:  
> 
> ```
> 
> consult-gh-emacs-pr-review.el:37:2: Error: Cannot open load file: No such file or directory, consult-gh
> 
> ```

I don't see this issue when byte-compiling. Is it possible that the dependency `consult-gh` is not loaded, when you byte-compiled?  

> 9:23: warning: This makes the package uninstallable in all released Emacs versions.  

In general the package should work on 29.4 or even earlier emacs versions but I am on emacs 30.0 and have not tested everything on 29.4 Some functions/features may have issues on 29.4 and earlier. I reduced the dependency to 29.4 now. Please advise if you think that's the right approach here or not.  

> 9:38: warning: Use a non-snapshot version number for dependency on "consult" if possible.  

consult recently had a breaking change (<https://github.com/minad/consult/blob/main/CHANGELOG.org>), but does not have a stable release tag with the new breaking change yet. The latest code in consult-gh has been adapted to be compatible with the latest consult, and to avoid non-compatibility I am pointing to a snapshot version until consult has a new stable release tag. Note that the stable release of consult-gh (v2.2) points to the compatible stabe version of consult (v1.9). Once consult has a new stable release, I can also make a new stable release tag and update the dependency to non-snapshot version.  

Let me know if that is the right approach or if you have some other suggestion on this.  

> 138:0: error: "consult-gh-topics-open-in-emacs-pr-review" doesn't start with package's prefix "consult-gh-emacs-pr-review".  

I think this should be ok in this case. This package is an extension for `consult-gh` so it should be ok to use function names that just have `consult-gh`. Note that this specific function is an interactive one available in `consult-gh-topics` buffers and following the same pattern as other `consult-gh-topics-*` functions would make it easier for users to find the right interactive command in the minibuffer.  

> ⸺ `consult-gh-emacs-pr-review.el` with [melpazoid](https://github.com/riscy/melpazoid):  
> 
> ```
> 
> - consult-gh-emacs-pr-review.el#L150: No `format` required; `message` takes an f-string
> 
> ```
> 
> ⸺ `consult-gh-emacs-pr-review.el` with checkdoc 0.6.2 (fix **within reason**):  
> 
> ```
> 
> consult-gh-emacs-pr-review.el:42: Lisp symbol `pr-review' should appear in quotes
> consult-gh-emacs-pr-review.el:52: Lisp symbol `pr-review' should appear in quotes
> consult-gh-emacs-pr-review.el:139: Lisp symbol `pr-review' should appear in quotes
> 
> ```

Those are taken care of now.

armindarvish commented:
> 9:38: warning: Use a non-snapshot version number for dependency on "consult" if possible.  
> 
> consult recently had a breaking change (<https://github.com/minad/consult/blob/main/CHANGELOG.org>), but does not have a stable release tag with the new breaking change yet. The latest code in consult-gh has been adapted to be compatible with the latest consult, and to avoid non-compatibility I am pointing to a snapshot version until consult has a new stable release tag. Note that the stable release of consult-gh (v2.2) points to the compatible stabe version of consult (v1.9). Once consult has a new stable release, I can also make a new stable release tag and update the dependency to non-snapshot version.  
> 
> Let me know if that is the right approach or if you have some other suggestion on this.  

consult has a new stable releas enow. so I have changed the dependency now.

riscy commented:
> consult has a new stable releas enow. so I have changed the dependency now.

Thanks for this - I guess this is probably the best outcome overall, but I do appreciate the time you took to explain the situation you faced to me.

Here's the latest -
⸺ `consult-gh-emacs-pr-review.el` with byte-compile using Emacs 29.4:
```
consult-gh-emacs-pr-review.el:37:2: Error: End of file while parsing JSON
```
You can see at the bottom of my comment that this appears to be because `gh` isn't installed on the testing system.  Since `consult-gh` is your own package, I think this is worth fixing upstream (a `user-error` would be much better than a user seeing a json parse error).  But - without jumping into the consult-gh source - I also question whether you really need to be calling out to `gh` and (presumbly?) creating a network connection when a user is simply loading in consult-gh.

⸺ `consult-gh-emacs-pr-review.el` with package-lint 20250108.1407:
```
2 issues found:
1:0: warning: The word "emacs" is redundant in Emacs package names.
138:0: error: "consult-gh-topics-open-in-emacs-pr-review" doesn't start with package's prefix "consult-gh-emacs-pr-review".
```
I think the first one is fine, but I still think the second one should be resolved.  I do see you wrote "This package is an extension for consult-gh so it should be ok to use function names that just have consult-gh" - but if we used that as a general rule we'd certainly start running into namespace collisions.  I do understand that you maintain consult-gh, so you would be able to avoid that, but I like general rules because they're far easier to enforce (or at least insist on :)

Since this package is `consult-gh-emacs-pr-review`, this function should be in the same namespace, e.g. named `consult-gh-emacs-pr-review-topics-open` or similar.

⸺ `consult-gh-emacs-pr-review.el` with checkdoc 0.6.2 (fix *within reason*):
```
consult-gh-emacs-pr-review.el:139: Lisp symbol `consult-gh' should appear in quotes
```

`#'load`-check on each file:
```
Loading consult-gh-emacs-pr-review.el
  consult-gh-emacs-pr-review.el:Error: Emacs 29.4:
  (json-end-of-file)
```

Stderr output while compiling/loading:
```
"gh" is not found on this system
"gh" is not found on this system
"gh" is not found on this system
```

armindarvish commented:
> ou can see at the bottom of my comment that this appears to be because `gh` isn't installed on the testing system.  Since `consult-gh` is your own package, I think this is worth fixing upstream (a `user-error` would be much better than a user seeing a json parse error).  But - without jumping into the consult-gh source - I also question whether you really need to be calling out to `gh` and (presumbly?) creating a network connection when a user is simply loading in consult-gh.  

Yes, I agree. I added a user-error and also removed the call to "gh" on loading the package. I think the call was to get the logged in username for `gh`, and cache it, instead we are now doing it the first time it is needed and caching it.  

Question for you. I am now throwing a `user-error`, if =gh = is not available per your suggestions. But, since I am also removing the call to gh on loading, I wonder if we should throw a user-error or just a warning?  

> Since this package is `consult-gh-emacs-pr-review`, this function should be in the same namespace, e.g. named `consult-gh-emacs-pr-review-topics-open` or similar.  

Ok, changed it.

riscy commented:
> I wonder if we should throw a user-error or just a warning?

It depends, but `user-error` will stop evaluation and interrupt whatever the user is doing.  (An `error` does this too but can open a backtrace, even when we know the error is on the user's side, not the elisp side.)

A `warn` will not stop evaluation and could lead to your program continuing on in an erroneous state.

Here's the latest -
⸺ `consult-gh-emacs-pr-review.el` with byte-compile using Emacs 29.4:
```
consult-gh-emacs-pr-review.el:37:2: Error: "gh" is not found on this system
```
It seems like `consult-gh` will throw byte-compile *errors* (i.e. it won't get byte compiled!) if the user doesn't have gh installed.  Are you sure this is what you want?  Perhaps you want them to be able to install consult-gh, byte-compile it, and then install `gh` after the fact.

⸺ `consult-gh-emacs-pr-review.el` with package-lint 20250108.1407:
```
1 issue found:
1:0: warning: The word "emacs" is redundant in Emacs package names.
```
I think here you're referring to the `pr-review` package when you say "Emacs pr-review" - I think saying something like `"pr-review" integration for consult-gh` makes it clear that you're referring to an Emacs package.

⸺ `consult-gh-emacs-pr-review.el` with checkdoc 0.6.2 (fix *within reason*):
```
consult-gh-emacs-pr-review.el:139: Lisp symbol `consult-gh' should appear in quotes
```
Minor!

`#'load`-check on each file:
```
Loading consult-gh-emacs-pr-review.el
  consult-gh-emacs-pr-review.el:Error: Emacs 29.4:
  (user-error #("\"gh\" is not found on this system" 0 32 (face warning)))
```
Just fyi.

armindarvish commented:
> ⸺ `consult-gh-emacs-pr-review.el` with byte-compile using Emacs 29.4:  
> 
> consult-gh-emacs-pr-review.el:37:2: Error: "gh" is not found on this system  
> 
> It seems like `consult-gh` will throw byte-compile **errors** (i.e. it won't get byte compiled!) if the user doesn't have gh installed.  Are you sure this is what you want?  Perhaps you want them to be able to install consult-gh, byte-compile it, and then install `gh` after the fact.  

This is why I was wondering if a warning would be a better approach rather than a user-error. This check/warning is currently happening at the very beginning of consult-gh.el, which I think is what preventing the byte compilation.  
So I changed the first check at the very top to a `warning` and elsewhere in the functions to a `user-error`. I don't have a system without gh to test it, so let me know how this behaves on the test system without gh now.  

> ⸺ `consult-gh-emacs-pr-review.el` with package-lint 20250108.1407:  
> 
> ```
> 
> 1 issue found:
> 1:0: warning: The word "emacs" is redundant in Emacs package names.
> 
> ```
> 
> I think here you're referring to the `pr-review` package when you say "Emacs pr-review" - I think saying something like `"pr-review" integration for consult-gh` makes it clear that you're referring to an Emacs package.  

Yes, that is correct. By using emacs in the name I am refering to integration with emacs `pr-review` package. I cannot just use `consult-gh-pr-review` as the name because that is already used elsewhere within consult-gh itself for some commands and it would become confusing. So I use `emacs-pr-review` here instead, but in all the docstrings I explicitly mention `'pr-review` and I also changed the header of the file with your suggestion as well. So I think everything should be clear here.  

> ⸺ `consult-gh-emacs-pr-review.el` with checkdoc 0.6.2 (fix **within reason**):  
> 
> ```
> 
> consult-gh-emacs-pr-review.el:139: Lisp symbol `consult-gh' should appear in quotes
> 
> ```
> 
> Minor!  

Fixed that.

riscy commented:
First, I think your approach with warnings looks good now.

But let's talk about the name, because I think `consult-gh-emacs-pr-review.el` is at least suboptimal.

> By using emacs in the name I am refering to integration with emacs pr-review package.

For my own bookkeeping here, you have a dependency on [pr-review](https://github.com/blahgeek/emacs-pr-review) (which happens to be in a repo called `emacs-pr-review`).

> I cannot just use consult-gh-pr-review as the name because that is already used elsewhere within consult-gh itself for some commands and it would become confusing.

I may be missing something but I see only [one isolated instance of consult-gh-pr-review](https://github.com/search?q=repo%3Aarmindarvish%2Fconsult-gh%20consult-gh-pr-review&type=code) using GitHub's code search and it's similar if I do a Google search (in a package named `consult-gh-embark` that should probably be calling that variable `consult-gh-embark-pr-review` per namespacing conventions).  And I also didn't realize that you basically maintain the entire [consult-gh](https://github.com/armindarvish/consult-gh) ecosystem.  Kudos for that, by the way - but my point is that you have a ton of control over the situation - this isn't us trying to accommodate someone else's package name.

armindarvish commented:
> For my own bookkeeping here, you have a dependency on [pr-review](https://github.com/blahgeek/emacs-pr-review) (which happens to be in a repo called `emacs-pr-review`).  

Yes, that is correct.  

> I may be missing something but I see only [one isolated instance of consult-gh-pr-review](https://github.com/search?q=repo%3Aarmindarvish%2Fconsult-gh%20consult-gh-pr-review&type=code) using GitHub's code search and it's similar if I do a Google search (in a package named `consult-gh-embark` that should probably be calling that variable `consult-gh-embark-pr-review` per namespacing conventions).  And I also didn't realize that you basically maintain the entire [consult-gh](https://github.com/armindarvish/consult-gh) ecosystem.  Kudos for that, by the way - but my point is that you have a ton of control over the situation - this isn't us trying to accommodate someone else's package name.  

Yes, I do have the control. But, if I just get rid of "emacs", then I end up with `consult-gh-pr-review.el` and need to name everything as `consult-gh-pr-review-*` by convention and for consistency. That will create an issue becuase I already have an interactive command called `consult-gh-pr-review` in consult-gh itself nad it will also be confusing to the user because the variables and funcitons in this function will all have the prefx consult-gh-pr-review, but there is no indicaiton that this is from the integration with another package and not built-in functionalities of consult-gh.  
So I want to use a different name thatn consult-gh-pr-review. Honestly, I don't understand what the problem with having "emacs" in the name is, so I thought that should be fine but I am open to other suggestions like consult-gh-pr-review-package-\*, or if you have any better suggestions to reflect the difference between features for reviewing a pr in consult-gh vs. features in this package that intergates the consult-gh with the `pr-review` package.

riscy commented:
> That will create an issue becuase I already have an interactive command called consult-gh-pr-review in consult-gh

I'd suggest you call this function `consult-gh-review-pr` (because that's what it's for, reviewing a PR) and then the deliberate namespace overlap with `pr-review` is entirely contained inside the package `consult-gh-pr-review`.

armindarvish commented:
> I'd suggest you call this function `consult-gh-review-pr`   

I don't like that idea because it breaks the logical consistency of the consult-gh commands. All commands follow the pattern of the gh command line. `gh search repos` becomes `consult-gh-search-repos`,  `gh pr review`  becomes `consult-gh-pr-review` and so on. I do not want to break that!

riscy commented:
> I don't like that idea because it breaks the logical consistency of the consult-gh commands.

OK, this is fair, and I hadn't considered that.  But I still think you can do better than `consult-gh-emacs-pr-review` if you are willing to compromise a bit and help me with my goals around package clarity.  Let me take another stab at it - instead of the ambiguous `emacs` in the middle of the package name, I think `consult-gh-with-pr-review` would make it clear to users that `pr-review` is the distinguishing feature being brought in.

armindarvish commented:
> I think `consult-gh-with-pr-review` would make it clear to users that `pr-review` is the distinguishing feature being brought in.  

Sounds good! I renamed the package to `consult-gh-with-pr-review` per your suggestion.

riscy commented:
Thanks you hugely for accommodating!  I'll merge this.

riscy commented:
One last note - In case you haven't, ensure GitHub release v2.2 is up-to-date with your current code and `Package-Version` (especially considering the name change).
